test/unit/puppetlabs/services/jruby/jruby_pool_test.clj

```diff
@@ -99,6 +99,62 @@
               (doseq [k (keys counts)]
                 (is (= (inc (counts k)) (new-counts k)))))))))))
 
+(deftest borrow-while-pool-is-being-initialized-test
+  (testing "borrow will block until an instance is available while the pool is coming online"
+    (let [pool-initialized? (promise)
+          init-fn (fn [instance] @pool-initialized? instance)
+          pool-size 1
+          config (jruby-testutils/jruby-config
+                  {:max-active-instances pool-size
+                   :lifecycle {:initialize-pool-instance init-fn}})
+          pool-context (jruby-core/create-pool-context config)]
+
+      ;; start a pool initialization, which will block on the `initialize-pool-instance`
+      ;; function's deref of the promise
+      (jruby-agents/send-prime-pool! pool-context)
+
+      ;; start a borrow, which should block until an instance becomes available
+      (let [borrow-instance (future (jruby-core/borrow-from-pool-with-timeout
+                                     pool-context
+                                     :borrow-during-pool-init-test
+                                     []))]
+
+        (is (not (realized? borrow-instance)))
+
+        ;; deliver the promise, allowing the pool initialization to complete
+        (deliver pool-initialized? true)
+
+        ;; now the borrow can complete
+        (is (some? @borrow-instance))))))
+
+(deftest borrow-while-no-instances-available-test
+  (testing "when all instances are in use, borrow blocks until an instance becomes available"
+    (let [pool-size 2
+          config (jruby-testutils/jruby-config
+                  {:max-active-instances pool-size})
+          pool-context (jruby-core/create-pool-context config)]
+      (jruby-agents/prime-pool! pool-context)
+
+      ;; borrow both instances from the pool
+      (let [drained-instances (jruby-testutils/drain-pool pool-context pool-size)]
+        (is (= 2 (count drained-instances)))
+
+        ;; attempt a borrow, which will block because no instances are free
+        (let [borrow-instance (future (jruby-core/borrow-from-pool-with-timeout
+                                       pool-context
+                                       :borrow-with-no-free-instances-test
+                                       []))]
+          (is (not (realized? borrow-instance)))
+
+          ;; return an instance to the pool
+          (jruby-core/return-to-pool
+           (first drained-instances)
+           :borrow-with-no-free-instances-test
+           [])
+
+          ;; now the borrow can complete
+          (is (some? @borrow-instance)))))))
```

Comment:
When the pool falls out of scope without being shut down, and without having all of its instances returned to it, what happens?  I feel like we had some tests in the past where not cleaning up the pool properly before letting the test exit meant that there were still threads doing things in the background after the test ran, and that could cause problems with subsequent tests.  I don't *think* that's the case here, but thought it was worth mentioning.  I'd probably consider it 'best practice' to do a blocking shutdown of any pool that you create in a test before exiting from the test block, but if there are a ton of tests that would need to be updated for that, I don't know that I'd block the PR on it.

I think we got away with this a little more easily with the old service because it had a `stop` lifecycle function that would automatically get called by TK when we exited the `with-app-with-config` block of the test, and that would shut the pool down.

Raises the question of whether or not the new pool manager service should have a `stop` lifecycle that tries to clean up pools that haven't been cleaned up by the time it runs... but that definitely seems like a separate chunk of work.